File: Server API - luis/probe-server/src/main/java/pt/solvit/probe/server/Interceptors/AuthInterceptor.java
package pt.solvit.probe.server.Interceptors;


import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.method.HandlerMethod;
import org.springframework.web.servlet.HandlerInterceptor;
import pt.solvit.probe.server.controller.exception.ForbiddenException;
import pt.solvit.probe.server.controller.exception.UnauthorizedException;
import pt.solvit.probe.server.model.User;
import pt.solvit.probe.server.service.api.IAuthenticationService;
import pt.solvit.probe.server.service.impl.UserService;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.time.LocalDateTime;
import java.util.Base64;

@Component
public class AuthInterceptor implements HandlerInterceptor {

    @Autowired
    IAuthenticationService authenticationService;


    @Override
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
        if(handler instanceof HandlerMethod /*&& ((HandlerMethod) handler).getMethod().getDeclaringClass().isAnnotationPresent(NoAuthorizationRequired.class)*/ ) {

            String authHeader = request.getHeader("Authorization");

            if (authHeader == null)
                throw new UnauthorizedException("Invalid token.", "Token is null.", "string", "about:blank");

            String[] authHeaderArr = authHeader.split(" ");
            if (authHeaderArr.length != 2 && !authHeaderArr[0].equals("Basic")) {
                throw new UnauthorizedException("Invalid token.", "Token is null or not Basic.", "string", "about:blank");
            }

            User user = verify(authHeaderArr[1]);

            request.setAttribute("user", user);
            /*
            request.setAttribute("userID",user.getId());
            request.setAttribute("userName",user.getUserName());
            request.setAttribute("userProfile",user.getUserProfile());//.toUpperCase());
            */

        }

        return true;
    }

    private User verify(String token){
        String[] decodedToken = new String(Base64.getDecoder().decode(token)).split(":");

        if(decodedToken.length != 2)
            throw new UnauthorizedException("Invalid token.", "Token is not Basic.", "string", "about:blank");

        User user = authenticationService.getAuthenticatedUser(decodedToken[0], decodedToken[1]);


        if (user.getSuspended())
            throw new ForbiddenException("User suspended.", "This user is suspended.", "string", "about:blank");

        return user;
    }
}

File: Server/MQSCF/src/main/kotlin/com/isel/project/MQSCF/config/AuthInterceptor.kt
package com.isel.project.MQSCF.config

import com.isel.project.MQSCF.model.Probeuser
import com.isel.project.MQSCF.utils.JsonProblemException
import org.springframework.stereotype.Component
import org.springframework.web.method.HandlerMethod
import org.springframework.web.servlet.HandlerInterceptor
import java.util.*
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class AuthInterceptor(val user : Probeuser) : HandlerInterceptor {


    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
        if(handler is HandlerMethod && handler.method.declaringClass.isAnnotationPresent(ProtectedRoute::class.java)) {
            val authHeader = request.getHeader("Authorization")?.split(" ")
            if (authHeader?.get(0) != "Basic") {
                throw JsonProblemException("You must provide a token to access that resource","token-error","no token",401,null,null)
            }
            val userID = verify(authHeader?.get(1))
            request.setAttribute("userID",userID)
        }

        return true
    }

    fun verify(token : String) =
            token
                .let {
                    String(Base64.getDecoder().decode(it)).split(":")
                }
                .also {
                    if(it.size != 2)
                        throw JsonProblemException("Token provided is not valid, Authentication must be Basic","token-error","Invalid token",401,null,null)
                }.let {
                        user.getAuthenticatedUser(it[0],it[1])
                    }
                    .let { it.id }

}